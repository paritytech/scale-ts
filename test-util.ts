/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
