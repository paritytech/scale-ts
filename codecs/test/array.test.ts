import * as $ from "../../mod.ts"
import { testCodec, testInvalid } from "../../test-util.ts"

testCodec($.array($.u8), [
  [1, 2, 3, 4, 5],
  [6, 7, 8, 9, 10],
  [11, 12, 13, 14, 15],
  [16, 17, 18, 19, 20, 21],
])

testInvalid($.array($.u8), [
  null,
  undefined,
  123,
  ["abc"],
  [1, 2, 3, -1, 4],
])

testCodec($.sizedArray($.u8, 1), [[1]])
testCodec($.sizedArray($.u8, 2), [[1, 1]])
testCodec($.sizedArray($.u8, 100), { "[1] * 100": Array(100).fill(1) as any })

testInvalid($.sizedArray($.u8, 3), [
  null,
  undefined,
  123,
  [],
  [1, 2, -1],
])

testCodec($.uint8Array, [
  new Uint8Array([1, 2, 3, 4, 5]),
  new Uint8Array([6, 7, 8, 9, 10]),
  new Uint8Array([11, 12, 13, 14, 15]),
  new Uint8Array([16, 17, 18, 19, 20, 21]),
])

testInvalid($.uint8Array, [
  null,
  undefined,
  123,
  [],
  [1, 2, -1],
])

testCodec($.sizedUint8Array(4), [
  new Uint8Array([0, 0, 0, 0]),
  new Uint8Array([1, 2, 3, 4]),
])

testInvalid($.sizedUint8Array(4), [
  null,
  undefined,
  [1, 2, 3, 4],
  new Uint8Array(),
  new Uint8Array(5),
])
