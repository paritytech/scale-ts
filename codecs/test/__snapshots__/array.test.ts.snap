export const snapshot = {};

snapshot[`\$.array(\$.u8) [ 1, 2, 3, 4, 5 ] 1`] = `
14
01
02
03
04
05
`;

snapshot[`\$.array(\$.u8) [ 6, 7, 8, 9, 10 ] 1`] = `
14
06
07
08
09
0a
`;

snapshot[`\$.array(\$.u8) [ 11, 12, 13, 14, 15 ] 1`] = `
14
0b
0c
0d
0e
0f
`;

snapshot[`\$.array(\$.u8) [ 16, 17, 18, 19, 20, 21 ] 1`] = `
18
10
11
12
13
14
15
`;

snapshot[`\$.sizedArray(\$.u8, 1) [ 1 ] 1`] = `01`;

snapshot[`\$.sizedArray(\$.u8, 2) [ 1, 1 ] 1`] = `
01
01
`;

snapshot[`\$.sizedArray(\$.u8, 100) [
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
] 1`] = `
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
`;

snapshot[`\$.uint8Array Uint8Array(5) [ 1, 2, 3, 4, 5 ] 1`] = `
14
01
02
03
04
05
`;

snapshot[`\$.uint8Array Uint8Array(5) [ 6, 7, 8, 9, 10 ] 1`] = `
14
06
07
08
09
0a
`;

snapshot[`\$.uint8Array Uint8Array(5) [ 11, 12, 13, 14, 15 ] 1`] = `
14
0b
0c
0d
0e
0f
`;

snapshot[`\$.uint8Array Uint8Array(6) [ 16, 17, 18, 19, 20, 21 ] 1`] = `
18
10
11
12
13
14
15
`;

snapshot[`\$.sizedUint8Array(4) Uint8Array(4) [ 0, 0, 0, 0 ] 1`] = `
00
00
00
00
`;

snapshot[`\$.sizedUint8Array(4) Uint8Array(4) [ 1, 2, 3, 4 ] 1`] = `
01
02
03
04
`;
